portfolio/src/main/java/com/google/sps/servlets/ServerLoginServlet.java

```diff
@@ -0,0 +1,39 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("login")
+public final class ServerLoginServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
+    PrintWriter out = response.getWriter();
+
+    UserService userService = UserServiceFactory.getUserService();
+
+    // If user is not logged in
+    if (!userService.isUserLoggedIn()) {
+      String loginUrl = userService.createLoginURL("/contact.html");
+      out.println("Log in <a href=\"" + loginUrl + "\">here</a> to add a comment.");
```

Comment:
Tip: Hyperlink **Log in** instead of "here". If the link text is descriptive, then it's much easier to guess what happens for when you click on the link as a user. For a more detailed discussion see https://www.w3.org/TR/WCAG10-HTML-TECHS/#link-text. Also think about accessibility here, what if you wanted a screen reader just read out all the links on the page to you?